File: Two_points/227.basic-calculator-ii.go
package leetcode

// @leet start

func getNum(s string, idx int) (num, step int) {
	for idx < len(s) && s[idx] >= '0' && s[idx] <= '9' {
		num *= 10
		num += int(s[idx] - '0')
		step++
		idx++
	}
	return num, step
}

func calculate(s string) int {
	var nums []int
	idx := 0
	ops := '.'
	for idx < len(s) {
		r := rune(s[idx])
		if r == ' ' {
			idx++
			continue
		}
		if r >= '0' && r <= '9' {
			curNum, step := getNum(s, idx)
			idx += step
			// handle number
			if ops == '.' {
				nums = append(nums, curNum)
				continue
			} else {
				// do operation
				switch ops {
				case '+':
					nums = append(nums, curNum)
				case '-':
					nums = append(nums, -curNum)
				case '*':
					lastItem := nums[len(nums)-1]
					nums = nums[:len(nums)-1]
					nums = append(nums, lastItem*curNum)
				case '/':
					lastItem := nums[len(nums)-1]
					nums = nums[:len(nums)-1]
					nums = append(nums, lastItem/curNum)
				}
				// reset ops
				ops = '.'
			}
		}
		ops = r
		idx++
	}

	var res int
	for _, v := range nums {
		res += v
	}
	return res
}

// @leet end

File: TwoPointer/227.basic-calculator-ii.py
# @leet start
class Solution:
    def calculate(self, s: str) -> int:
        ret = prev = num = 0
        op = "+"
        for c in s + "+":
            if c == " ":
                continue
            if c.isdigit():
                num = num * 10 + int(c)
                continue
            match op:
                case "+":
                    prev = num
                case "-":
                    prev = -num
                case "*":
                    ret -= prev
                    prev *= num
                case "/":
                    ret -= prev
                    prev = int(prev / num)
            ret += prev
            op, num = c, 0
        return ret


# @leet end
